pkg/apis/discovery/types.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// EndpointSlice represents a subset of the endpoints that implement a service.
+// For a given service there may be multiple EndpointSlice objects, selected by
+// labels, which must be joined to produce the full set of endpoints.
+type EndpointSlice struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// +optional
+	metav1.ObjectMeta
+	// endpoints is a list of unique endpoint objects in this slice.
+	// +listType=atomic
+	Endpoints []Endpoint
+	// ports specifies the list of network ports exposed each endpoint in this
+	// EndpointSlice. Each port must have a unique name. When endpoints is nil,
+	// it indicates that ports are irrelevant for the address type. When this is
+	// an empty slice, it indicates that ports are relevant for the address type,
+	// but none are currently relevant. The most common example of this occurs
+	// when no pods match a Service selector and a placeholder EndpointSlice is
+	// generated.
+	// +optional
+	// +listType=atomic
```

Comment:
do we want to specify a limit to number of ports?  100?  200?  500?

Reply:
This needs to match services though. Without port range, I would imagine many people specifying many ports.

Reply:
There's a practical limit that we want to defend, and if it overflows it could just become a new slice, right?  E.g. if we limit to 100 ports and I need 101, make 2 slices? Even if we start at 500, we'll need some overflow handling... (and tests)

Reply:
I added an initial limit of 100 here, easy to change the value if another number makes more sense.